### ISSUE TITLE
-dsource dumps invalid program

### ISSUE THREAD

**Issue by bobzhang** [ocaml/ocaml#10623]
version 4.10.2

Input:
```
let rec pp_ident :
          'loc .
           'loc
              -> 'b

  =
  ((
      fun poly_loc ->
        x)
  [@ocaml.warning "-A"]
  )
and show_ident :
  'loc .
    'a
  =
  fun poly_loc ->
    x

```

Output:
```
let rec (pp_ident : 'loc . 'loc -> 'b) = ((fun poly_loc -> x)
  [@ocaml.warning "-A"])
and show_ident : 'loc . 'a = fun poly_loc -> x
```

The parens around `pp_ident` produces invalid program

**Comment by gasche** [comment]
I cannot reproduce with 4.12:

```
$ ocamlc -stop-after parsing -dsource test.ml
let rec pp_ident : 'loc . 'loc -> 'b = ((fun poly_loc -> x)
  [@ocaml.warning "-A"])
and show_ident : 'loc . 'a = fun poly_loc -> x
```

**Comment by gasche** [comment]
This was #9778, first merged in 4.12.
